### PR TITLE
drivers: display: st7789v: add rotation feature

### PR DESCRIPTION
**Pull request by cvetaevvitaliy**
Adds rotation function for st7789v display driver
Opened issue #[32286](https://github.com/zephyrproject-rtos/zephyr/issues/32286)

Signed-off-by: Vitaliy Nimych <vitaliy.nimych@gmail.com.com>

### CODE REVIEW

File: drivers/display/display_st7789v.c [hunk 1]
```diff
@@ -254,11 +255,48 @@ static int st7789v_set_pixel_format(const struct device *dev,
...
+
+	switch (orientation)
+	{
+	case DISPLAY_ORIENTATION_NORMAL:
+		tx_data |= ST7789V_MADCTL_MV_NORMAL_MODE;
+		x_offset = data->x_offset;
+		y_offset = data->y_offset;
+		break;
+
+	case DISPLAY_ORIENTATION_ROTATED_90:
```
Comment (caspermeijn) [inline drivers/display/display_st7789v.c:271]: My display is rotated counter clockwise 90 degrees. It is not documented which direction the rotation is expected. If feels to me more logical to do clockwise. Do you know what is the correct direction?

File: drivers/display/display_st7789v.c [hunk 2]
```diff
@@ -254,11 +255,48 @@ static int st7789v_set_pixel_format(const struct device *dev,
...
+		tx_data |= (ST7789V_MADCTL_MY_BOTTOM_TO_TOP | ST7789V_MADCTL_MV_REVERSE_MODE);
+		x_offset = data->y_offset;
+		y_offset = data->x_offset;
+		break;
+
+	case DISPLAY_ORIENTATION_ROTATED_180:
+		tx_data |= (ST7789V_MADCTL_MY_BOTTOM_TO_TOP | ST7789V_MADCTL_MX_RIGHT_TO_LEFT);
+		x_offset = data->x_offset;
+		y_offset = data->y_offset;
+		break;
```
Comment (caspermeijn) [inline drivers/display/display_st7789v.c:281]:
```suggestion
	case DISPLAY_ORIENTATION_ROTATED_90:
		tx_data |= (ST7789V_MADCTL_MY_BOTTOM_TO_TOP | ST7789V_MADCTL_MV_REVERSE_MODE);
		x_offset = 320 - data->height + data->y_offset;
		y_offset = 240 - data->width + data->x_offset;
		break;

	case DISPLAY_ORIENTATION_ROTATED_180:
		tx_data |= (ST7789V_MADCTL_MY_BOTTOM_TO_TOP | ST7789V_MADCTL_MX_RIGHT_TO_LEFT);
		x_offset = 240 - data->width + data->x_offset;
		y_offset = 320 - data->height + data->y_offset;
		break;
```
My display is 240x240. I need these extra constants to align the picture with the display.

**Review by caspermeijn — CHANGES_REQUESTED** [review]
Tested with PineTime with 240x240 display. It worked after correcting the offset (see other comment).

- You should squash the two commits into one.
- You should rebase the branch to the latest main, as some changes are made to the st7789v driver.

**Comment by cvetaevvitaliy** [comment]
> Tested with PineTime with 240x240 display. It worked after correcting the offset (see other comment).
> 
> 
> 
> - You should squash the two commits into one.
> 
> - You should rebase the branch to the latest main, as some changes are made to the st7789v driver.

Hi!
Yes, I received yesterday one display 240x240 for test

I want to finish my work, on New Year's Eve and start testing driver for 240x240 resolution.

Thanks for answer and help with testing.

**Comment by JON95Git** [comment]
> > Tested with PineTime with 240x240 display. It worked after correcting the offset (see other comment).
> > 
> > * You should squash the two commits into one.
> > * You should rebase the branch to the latest main, as some changes are made to the st7789v driver.
> 
> Hi! Yes, I received yesterday one display 240x240 for test
> 
> I want to finish my work, on New Year's Eve and start testing driver for 240x240 resolution.
> 
> Thanks for answer and help with testing.
> 
> image

Hello @cvetaevvitaliy. Can you perform the tests with this generic 240x240 display? I'm trying to test on the same display version (https://www.aliexpress.com/item/32915085705.html) and it doesn't work.

**Comment by ck-telecom** [comment]
@cvetaevvitaliy please check my comments, and I tried and it is ok for my 240x240 lcd, it is ok!

**Comment by ck-telecom** [comment]
@cvetaevvitaliy If you still want to continue this PR, please
- resolve the review issue
- squash the two commits into one
- rebase the branch to the latest main, as some changes are made to the st7789v driver

**Comment by cvetaevvitaliy** [comment]
> If you still want to continue this PR, please
> 
> resolve the review issue
> squash the two commits into one
> rebase the branch to the latest main, as some changes are made to the st7789v drive

I will try to do it soon.  I am in Ukraine, it was hot for a couple of months

**Pull request closed without merging**